src/domain_extractor.coffee

```diff
@@ -0,0 +1,111 @@
+define [
+  'settings'
+], (Settings)->
+  ###
+    First, we consider the part after the last dot as part of the TLD.
+    Then, starting from right to left we consider anything that is less than
+    4 characters long as part of the TLD. When we reach a level that is
+    at least 4 characters we include it in the domain and we stop.
+
+    We also accept a custom TLD list that allows us to bypass this.
+    In that case, the domain is determined as the custom TLD plus
+    the part immediately preceding it.
+
+    Finally, we always drop the www prefix.
+
+    This approach might be problematic in the future, but the only safe
+    alternative is to use the Public Suffix List which contains all
+    the known TLDs. However is quite large in size.
+
+    @see https://github.com/wrangr/psl
+  ###
+  class DomainExtractor
+    NUMBER_DOT_ONLY_PATTERN = /^[\d.]+$/
+
+    constructor: (hostname) ->
+      @hostname = hostname.toLowerCase();
+
+    ###
+      Retrieve the base domain from the given hostname
+
+      @param [Boolean] wildcard Whether to append a dot in the beginning of the domain.
+      @return [String, null] The base domain or null if is IP or 1 level
+    ###
+    get: (wildcard = true) ->
+      return null if @hostname.match NUMBER_DOT_ONLY_PATTERN # check if it is IP
+
+      @domainParts = @hostname.split('.');
+      @domainParts.shift() if @domainParts[0] == 'www' # remove the www prefix
+      return null if @domainParts.length == 1 # check if it is 1 level
```

Comment:
What about this one `www.shop`? 
:rofl:

Reply:
fixed 😠